daml_dit_if/main/auth_handler.py

```diff
@@ -0,0 +1,150 @@
+    """
+    Mark a request handler as not requiring authentication.
+    """
+    setattr(fn, DABL_AUTH_LEVEL, auth)
+
+    return fn
+
+
+def auth_level(auth: "AuthorizationLevel") -> "Callable[[Handler], Handler]":
+
+    def set(fn: "Handler") -> "Handler":
+        return set_handler_auth(fn, auth)
+
+    return set
+
+
+def get_handler_auth_level(request: "Request") -> 'AuthorizationLevel':
+    return getattr(request.match_info.handler, DABL_AUTH_LEVEL, AuthorizationLevel.PUBLIC)
+
+
+def get_ledger_claims(
+        config: 'Configuration', claims: "Mapping[str, Any]") -> "Optional[Mapping[str, Any]]":
+
+    ledger_claims = claims.get('https://daml.com/ledger-api')
+
+    LOG.debug('ledger_claims: %r', ledger_claims)
+
+    if ledger_claims is None:
+        return None
+
+    claimed_ledger_id = ledger_claims.get('ledgerId', 'missing-ledger-id-claim')
+
+    if claimed_ledger_id != config.ledger_id:
+        LOG.debug(f'Ledger ID mismatch in claims: {claimed_ledger_id} != {config.ledger_id}')
+        return None
+
+    return ledger_claims
+
+
+def is_integration_party_ledger_claim(config: "Configuration", ledger_claims: "Mapping[str, Any]") -> bool:
+    act_as_parties = ledger_claims.get('actAs', [])
+    read_as_parties = ledger_claims.get('readAs', [])
+
+    party = config.run_as_party
+
+    if party is None:
+        return False
+
+    return party in act_as_parties and party in read_as_parties
+
+
+class AuthHandler:
+    def __init__(self, config: 'Configuration', jwt_decoder: 'Optional[JWTValidator]'):
+        self.config = config
+        self.jwt_decoder = jwt_decoder
+
+    async def setup(self, app: "Application") -> None:
+        app.middlewares.append(self.auth_middleware)
+
+    @middleware
+    async def auth_middleware(self, request: "Request", handler):
+        LOG.debug("in auth middleware for request %s", request)
+
+        auth_level = get_handler_auth_level(request)
+
+        if auth_level != AuthorizationLevel.PUBLIC:
+
+            if self.jwt_decoder is None:
+                raise unauthorized_response(
+                    "no_authorization_support",
+                    "this endpoint requires authorization, which is unavailable without JWKS support.",
+                )
+
+            token = get_token(request)
+            if token is None:
+                raise unauthorized_response(
+                    "missing_token",
+                    "this endpoint requires a valid token and none was supplied",
+                )
+
+            try:
+                claims = await self.jwt_decoder.decode_claims(token)
+            except JWException as ex:
+                LOG.warning("Rejected a token: %s", ex)
+                raise forbidden_response(
+                    "invalid_token", "this endpoint was presented with an invalid token"
+                )
+
+            ledger_claims = get_ledger_claims(self.config, claims)
```

Comment:
it wasn't clear to me that this primarily checks ledger id - could this be renamed to more precisely reflect that it asserts a ledger id match etc?

Reply:
Sure... the intent of the existing naming is that it gets the claims for the _specific configured ledger_... does that make it clearer? Assuming no, since that line of thinking wasn't initially obvious

Reply:
oh i see. i thought `ledger` was referring to the ledger claims section of the token itself, and not the `ledger` upon which the integration was configured

Reply:
I'm preparing another PR today that will touch this. Will review the naming as part of that.